File: backend/app/schemas/legal.py
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any, Literal

from pydantic import BaseModel, Field


class LocalizedText(BaseModel):
    pt: str = Field(default="")
    es: str = Field(default="")


class LegalVariableDefinition(BaseModel):
    key: str
    placeholder: str
    label: LocalizedText
    description: LocalizedText
    sample_value: str


class LegalVariableCategory(BaseModel):
    key: str
    label: LocalizedText
    variables: list[LegalVariableDefinition]


class LegalVariablesResponse(BaseModel):
    categories: list[LegalVariableCategory]


class LegalTemplatePayload(BaseModel):
    name: str = Field(..., max_length=255)
    description: str | None = Field(default=None, max_length=500)
    content: str
    is_active: bool = True


class LegalTemplateSummary(BaseModel):
    id: int
    name: str
    description: str | None = None
    is_active: bool
    created_at: datetime
    updated_at: datetime | None = None


class LegalTemplateDetail(LegalTemplateSummary):
    content: str


class LegalTemplateListResponse(BaseModel):
    items: list[LegalTemplateSummary]


class LegalContractOut(BaseModel):
    id: int
    sale_id: int
    template_id: int | None = None
    buyer_name: str
    product_name: str
    total_amount: int
    currency: str
    status: str
    pdf_url: str | None = None
    created_at: datetime
    generated_at: datetime | None = None
    last_error: str | None = None
    signature_status: str
    signature_token: str | None = None
    signature_link: str | None = None
    signature_signed_at: datetime | None = None
    signature_name: str | None = None
    signature_type: str | None = None
    signature_image_url: str | None = None
    signature_ip: str | None = None
    signature_user_agent: str | None = None
    signed_pdf_url: str | None = None
    signed_pdf_generated_at: datetime | None = None
    signed_pdf_size_bytes: int | None = None
    document_hash: str | None = None
    document_hash_algorithm: str | None = None
    verification_token: str | None = None
    verification_url: str | None = None
    verification_qr_image_data: str | None = None
    verification_generated_at: datetime | None = None
    agency_signature_status: str
    agency_signature_signed_at: datetime | None = None
    agency_signature_type: str | None = None
    agency_signature_name: str | None = None
    agency_signature_role: str | None = None
    agency_signature_company: str | None = None
    agency_signature_city: str | None = None
    agency_signature_font_style: str | None = None
    agency_signature_typed_value: str | None = None
    agency_signature_image_url: str | None = None


class LegalContractListResponse(BaseModel):
    items: list[LegalContractOut]


class LegalContractDetail(LegalContractOut):
    pass


class LegalContractSignaturePublic(BaseModel):
    contract_id: int
    sale_id: int
    token: str
    buyer_name: str
    product_name: str
    currency: str
    total_amount: int
    status: str
    pdf_url: str | None = None
    signature_status: str
    signature_signed_at: datetime | None = None
    signature_name: str | None = None
    signature_type: str | None = None
    signature_image_url: str | None = None
    agency_name: str | None = None
    agency_logo_url: str | None = None
    created_at: datetime
    generated_at: datetime | None = None
    signed_pdf_url: str | None = None
    signed_pdf_generated_at: datetime | None = None
    signed_pdf_size_bytes: int | None = None
    document_hash: str | None = None
    document_hash_algorithm: str | None = None
    verification_token: str | None = None
    verification_url: str | None = None
    verification_qr_image_data: str | None = None
    verification_generated_at: datetime | None = None
    agency_signature_status: str
    agency_signature_signed_at: datetime | None = None
    agency_signature_type: str | None = None
    agency_signature_name: str | None = None
    agency_signature_role: str | None = None
    agency_signature_company: str | None = None
    agency_signature_city: str | None = None
    agency_signature_font_style: str | None = None
    agency_signature_typed_value: str | None = None
    agency_signature_image_url: str | None = None


class LegalContractSignatureSubmitPayload(BaseModel):
    signature_type: Literal["typed", "drawn"]
    full_name: str | None = None
    signature_image: str | None = None
    accepted_terms: bool = False


class LegalContractSignatureSubmitResponse(BaseModel):
    detail: LegalContractSignaturePublic


class LegalContractVerificationStatus(str, Enum):
    valid = "valid"
    pending = "pending"
    incomplete = "incomplete"
    invalid = "invalid"
    not_found = "not_found"


class LegalContractVerificationDetail(BaseModel):
    status: LegalContractVerificationStatus
    contract_id: int | None = None
    sale_id: int | None = None
    buyer_name: str | None = None
    product_name: str | None = None
    agency_name: str | None = None
    agency_logo_url: str | None = None
    created_at: datetime | None = None
    generated_at: datetime | None = None
    signature_signed_at: datetime | None = None
    signature_status: str | None = None
    signature_type: str | None = None
    agency_signature_status: str | None = None
    agency_signature_signed_at: datetime | None = None
    document_hash: str | None = None
    document_hash_algorithm: str | None = None
    signed_pdf_url: str | None = None
    signed_pdf_generated_at: datetime | None = None
    signed_pdf_size_bytes: int | None = None
    pdf_url: str | None = None
    verification_token: str | None = None
    verification_url: str | None = None
    verification_qr_image_data: str | None = None
    verification_generated_at: datetime | None = None
    message: str


class LegalContractAuditEventOut(BaseModel):
    id: int
    contract_id: int
    event_type: str
    title: str
    description: str | None = None
    actor_type: str
    occurred_at: datetime
    is_reconstructed: bool = False
    metadata: dict[str, Any] | None = None


class LegalContractAuditEventList(BaseModel):
    items: list[LegalContractAuditEventOut]
    has_more: bool = False


class LegalSignatureProfilePayload(BaseModel):
    signature_type: Literal["drawn", "typed"]
    signature_drawn_image: str | None = None
    signature_typed_value: str | None = None
    signature_font_style: str | None = None
    signature_display_name: str
    signature_role: str | None = None
    signature_company_name: str | None = None
    signature_city: str | None = None


class LegalSignatureProfileResponse(BaseModel):
    signature_type: str
    signature_drawn_image_url: str | None = None
    signature_drawn_image_data: str | None = None
    signature_typed_value: str | None = None
    signature_font_style: str | None = None
    signature_display_name: str
    signature_role: str | None = None
    signature_company_name: str | None = None
    signature_city: str | None = None
    updated_at: datetime | None = None

File: frontend/src/types/legal.ts
export interface LocalizedText {
  pt?: string;
  es?: string;
  [key: string]: string | undefined;
}

export interface LegalVariableDefinition {
  key: string;
  placeholder: string;
  label: LocalizedText;
  description: LocalizedText;
  sample_value: string;
}

export interface LegalVariableCategory {
  key: string;
  label: LocalizedText;
  variables: LegalVariableDefinition[];
}

export interface LegalVariablesResponse {
  categories: LegalVariableCategory[];
}

export interface LegalTemplateSummary {
  id: number;
  name: string;
  description?: string | null;
  is_active: boolean;
  created_at: string;
  updated_at?: string | null;
}

export interface LegalTemplateDetail extends LegalTemplateSummary {
  content: string;
}

export interface LegalTemplateListResponse {
  items: LegalTemplateSummary[];
}

export interface LegalTemplatePayload {
  name: string;
  description?: string | null;
  content: string;
  is_active: boolean;
}

export interface LegalContract {
  id: number;
  sale_id: number;
  template_id?: number | null;
  buyer_name: string;
  product_name: string;
  total_amount: number;
  currency: string;
  status: string;
  pdf_url?: string | null;
  created_at: string;
  generated_at?: string | null;
  last_error?: string | null;
  signature_status: string;
  signature_token?: string | null;
  signature_link?: string | null;
  signature_signed_at?: string | null;
  signature_name?: string | null;
  signature_type?: string | null;
  signature_image_url?: string | null;
  signature_ip?: string | null;
  signature_user_agent?: string | null;
  signed_pdf_url?: string | null;
  signed_pdf_generated_at?: string | null;
  signed_pdf_size_bytes?: number | null;
  document_hash?: string | null;
  document_hash_algorithm?: string | null;
  verification_token?: string | null;
  verification_url?: string | null;
  verification_qr_image_data?: string | null;
  verification_generated_at?: string | null;
  agency_signature_status: string;
  agency_signature_signed_at?: string | null;
  agency_signature_type?: string | null;
  agency_signature_name?: string | null;
  agency_signature_role?: string | null;
  agency_signature_company?: string | null;
  agency_signature_city?: string | null;
  agency_signature_font_style?: string | null;
  agency_signature_typed_value?: string | null;
  agency_signature_image_url?: string | null;
}

export interface LegalContractListResponse {
  items: LegalContract[];
}

export interface LegalContractSignaturePublic {
  contract_id: number;
  sale_id: number;
  token: string;
  buyer_name: string;
  product_name: string;
  currency: string;
  total_amount: number;
  status: string;
  pdf_url?: string | null;
  signature_status: string;
  signature_signed_at?: string | null;
  signature_name?: string | null;
  signature_type?: string | null;
  signature_image_url?: string | null;
  agency_name?: string | null;
  agency_logo_url?: string | null;
  created_at: string;
  generated_at?: string | null;
  signed_pdf_url?: string | null;
  signed_pdf_generated_at?: string | null;
  signed_pdf_size_bytes?: number | null;
  document_hash?: string | null;
  document_hash_algorithm?: string | null;
  verification_token?: string | null;
  verification_url?: string | null;
  verification_qr_image_data?: string | null;
  verification_generated_at?: string | null;
  agency_signature_status: string;
  agency_signature_signed_at?: string | null;
  agency_signature_type?: string | null;
  agency_signature_name?: string | null;
  agency_signature_role?: string | null;
  agency_signature_company?: string | null;
  agency_signature_city?: string | null;
  agency_signature_font_style?: string | null;
  agency_signature_typed_value?: string | null;
  agency_signature_image_url?: string | null;
}

export interface LegalContractSignatureSubmitPayload {
  signature_type: "typed" | "drawn";
  full_name?: string | null;
  signature_image?: string | null;
  accepted_terms: boolean;
}

export interface LegalContractSignatureSubmitResponse {
  detail: LegalContractSignaturePublic;
}

export type LegalContractVerificationStatus = "valid" | "pending" | "incomplete" | "invalid" | "not_found";

export interface LegalContractVerificationDetail {
  status: LegalContractVerificationStatus;
  contract_id?: number | null;
  sale_id?: number | null;
  buyer_name?: string | null;
  product_name?: string | null;
  agency_name?: string | null;
  agency_logo_url?: string | null;
  created_at?: string | null;
  generated_at?: string | null;
  signature_signed_at?: string | null;
  signature_status?: string | null;
  signature_type?: string | null;
  agency_signature_status?: string | null;
  agency_signature_signed_at?: string | null;
  document_hash?: string | null;
  document_hash_algorithm?: string | null;
  signed_pdf_url?: string | null;
  signed_pdf_generated_at?: string | null;
  signed_pdf_size_bytes?: number | null;
  pdf_url?: string | null;
  verification_token?: string | null;
  verification_url?: string | null;
  verification_qr_image_data?: string | null;
  verification_generated_at?: string | null;
  message: string;
}

export type SignatureFontStyle = "classic" | "cursive" | "elegant";

export interface LegalSignatureProfilePayload {
  signature_type: "drawn" | "typed";
  signature_drawn_image?: string | null;
  signature_typed_value?: string | null;
  signature_font_style?: SignatureFontStyle | null;
  signature_display_name: string;
  signature_role?: string | null;
  signature_company_name?: string | null;
  signature_city?: string | null;
}

export interface LegalSignatureProfile extends Omit<LegalSignatureProfilePayload, "signature_drawn_image"> {
  signature_drawn_image_url?: string | null;
  signature_drawn_image_data?: string | null;
  updated_at?: string | null;
}
